api/ai/new/geg_algorithm/geg_algorithm.py

```diff
@@ -0,0 +1,352 @@
+import math
+import random
+from typing import List, Dict, Tuple, Set
+
+from api.ai.new.interfaces.algorithm import Algorithm
+from api.ai.new.interfaces.algorithm_options import GeneralizedEnvyGraphAlgorithmOptions
+from api.ai.new.interfaces.team_generation_options import TeamGenerationOptions
+from api.models.enums import (
+    AlgorithmType,
+    ScenarioAttribute,
+    Gender,
+    Race,
+    RequirementOperator,
+)
+from api.models.project import Project, ProjectRequirement
+from api.models.student import Student
+from api.models.team import Team
+from api.models.team_set import TeamSet
+from benchmarking.data.simulated_data.mock_student_provider import (
+    MockStudentProviderSettings,
+    MockStudentProvider,
+)
+from benchmarking.evaluations.graphing.graph_metadata import GraphData
+
+
+def get_additive_utilities(
+    students: List[Student], projects: List[Project]
+) -> Dict[Tuple[int, int], int]:
+    utilities: Dict[Tuple[int, int], int] = {}
+
+    for project in projects:
+        project_requirements = set(project.requirements)
+        for student in students:
+            student_utilities = 0
+            student_attributes = student.attributes
+            for attribute_id, attribute_values in student_attributes.items():
+                if attribute_id in project_requirements:
+                    student_utilities += sum(
+                        [
+                            1
+                            for attribute_value in attribute_values
+                            if attribute_value > 0
+                        ]
+                    )
+                else:
+                    student_utilities -= 1
+
+            utilities[(project.id, student.id)] = student_utilities
+
+    return utilities
+
+
+class EnvyGraph:
```

Comment:
each algorithm gets its own directory so you really don't have to crunch every class and util function into this one file, it feels kinda messy to me